File: 416_partition_equal_subset_sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;

        for(int num: nums) sum += num;
        if((sum & 1) != 0) return false;

        sum /= 2;

        boolean[] dp = new boolean[sum+1];
        dp[0] = true;

        for(int num: nums){
            for(int i = sum; i > 0; i--){
                if(num <= i) dp[i] = dp[i] || dp[i-num];
            }
        }

        return dp[sum];
    }
}

File: 416_partition_equal_subset_sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if (total % 2) != 0:
            return False

        target = total//2

        # find subset that sums to target
        dp = [False] * (target+1)
        dp[0] = True

        for num in nums:
            for i in range(target, 0, -1):
                if i >= num: dp[i] = dp[i-num] or dp[i]
        
        return dp[-1]
        
